### PR TITLE
Code 204

### PR DESCRIPTION
**Pull request by ArBridgeman**

### CODE REVIEW

File: sous-chef/sous_chef/grocery_list/generate_grocery_list.py [hunk 1]
```diff
@@ -100,30 +103,26 @@ def upload_grocery_list_to_todoist(self, todoist_helper: TodoistHelper):
...
-        due_date = self.due_date_formatter.get_due_datetime_with_hour_minute(
-            weekday=bean_prep.prep_day, hour=bean_prep.prep_hour, minute=0
-        )
-
-        # TODO change queue_bean_preparation into dataframe to use pandas here
-        if self.queue_bean_preparation:
-            for item in self.queue_bean_preparation:
-                label_list = item["group"].from_recipe + item["group"].from_day
+        if self.queue_preparation:
+            for _, row in self.queue_preparation.iterrows():
```
Comment (ArBridgeman) [inline sous-chef/sous_chef/grocery_list/generate_grocery_list.py:116]: Should add tests for adding to queue_preparation here

File: sous-chef/sous_chef/grocery_list/generate_grocery_list.py [hunk 2]
```diff
@@ -202,18 +201,101 @@ def _add_menu_recipe_to_queue(self, menu_recipe_list: list[MenuRecipe]):
...
+                from_recipe=from_recipe,
                 from_day=menu_recipe.from_day,
                 eat_factor=menu_recipe.eat_factor * recipe.factor,
                 freeze_factor=menu_recipe.freeze_factor * recipe.factor,
                 recipe=recipe,
             )
-            self._add_menu_recipe_to_queue([menu_recipe])
+
+            # TODO make more robust to other methods
+            if self.config.run_mode.with_todoist:
```
Comment (ArBridgeman) [inline sous-chef/sous_chef/grocery_list/generate_grocery_list.py:271]: Add tests to cover all this new logic

**Review by AkagiShigeru — APPROVED** [review]
changes reviewed, found fine, tests and code working.

File: sous-chef/sous_chef/grocery_list/generate_grocery_list.py (hunk 1 above)
Comment (AkagiShigeru) [reply sous-chef/sous_chef/grocery_list/generate_grocery_list.py:119]: added, please check

**Review by AkagiShigeru — APPROVED** [review]
All recent changes reviewed.

File: sous-chef/sous_chef/grocery_list/generate_grocery_list.py (hunk 2 above)
Comment (AkagiShigeru) [reply sous-chef/sous_chef/grocery_list/generate_grocery_list.py:278]: could do that, not sure how critical. what do you think?

File: sous-chef/sous_chef/messaging/todoist_api.py [hunk 3]
```diff
@@ -33,10 +33,11 @@ class TodoistHelper:
     config: DictConfig
     connection: TodoistAPI = field(init=False)
     projects: Dict[str, Project] = field(init=False)
+    dry_run: bool = False
```
Comment (ArBridgeman) [inline sous-chef/sous_chef/messaging/todoist_api.py:36]: Why do you want to add this here?

**Comment by ArBridgeman** [comment]
Not stable to run some todoist api tests on github as frequently need to update api key
Real coverage for that is 94%.

**Pull request merged**